Razor syntax updates

guardrex opened the pull request:
Fixes #13465

[Internal Review Topic (Razor syntax)](https://review.docs.microsoft.com/en-us/aspnet/core/mvc/views/razor?view=aspnetcore-3.0&branch=pr-en-us-13524)
[Internal Review Topic (Components)](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/components?view=aspnetcore-3.0&branch=pr-en-us-13524)

* Hopefully not as stinky as that steaming pile 'o 🐄💩 that I turned in last time. :smile:
  * Need feedback where the scenario only applies to Razor components, but I don't say so. 🙈
  * Need feedback where the scenario applies to both RP/views and Razor components but I say it only applies to Razor components. 🙉
* Get away from `#section-{x}` bookmarks by escaping the `@` symbol in section headings. Now, we can cross-link properly. For example, the section heading `### \@using` can be cross-linked as `<xref:mvc/views/razor#using>`. Much better than `### @using` (unescaped `@`) where we would end up with something like `<xref:mvc/views/razor#section-7>` and then we can't reorganize or insert new sections without breaking the *numbered-by-DocFX* section link.
* Alphabetize the directives.
* Adds a new *Directive attributes* section.
* Engineering guidance (Ryan): `@functions`/`@code` aren't "code blocks." That language should be reserved for `@{...}` constructs.

NTaylorMullen commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -512,92 +585,279 @@ If "rick@contoso.com" is passed in the model, the view generates the following H
+### \@ref
 
-### @section
+Razor component references (`@ref`) provide a way to reference a component instance so that you can issue commands to that instance, such as `Show` or `Reset`. For more information, see <xref:blazor/components#capture-references-to-components>.
```
Might be valuable to include the note form this section: https://docs.microsoft.com/en-us/aspnet/core/blazor/components?view=aspnetcore-3.0#capture-references-to-components

NTaylorMullen commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -704,9 +964,9 @@ There are three directives that pertain to [Tag Helpers](xref:mvc/views/tag-help
-| [&commat;addTagHelper](xref:mvc/views/tag-helpers/intro#add-helper-label) | Makes Tag Helpers available to a view. |
-| [&commat;removeTagHelper](xref:mvc/views/tag-helpers/intro#remove-razor-directives-label) | Removes Tag Helpers previously added from a view. |
-| [&commat;tagHelperPrefix](xref:mvc/views/tag-helpers/intro#prefix-razor-directives-label) | Specifies a tag prefix to enable Tag Helper support and to make Tag Helper usage explicit. |
+| [\@addTagHelper](xref:mvc/views/tag-helpers/intro#add-helper-label) | Makes Tag Helpers available to a view. |
```
All these are only valid in Razor views (TagHelpers aren't a thing in Blazor files)

NTaylorMullen commented on aspnetcore/mvc/views/razor.md:
````diff
@@ -442,43 +442,113 @@ public class _Views_Something_cshtml : RazorPage<dynamic>
-```csharp
-public class _Views_Account_Login_cshtml : RazorPage<LoginViewModel>
+> [!NOTE]
+> In [Razor components](xref:blazor/components), use `@code` over `@functions` to add C# members.
````
In Blazor we should always recommend @code. Outside of Blazor users only have @functions

guardrex commented:
Thanks for reviewing this so quickly ...

> ... the Razor syntax reference doc should remain a syntax reference, not an entry point for learning the language. You come to this doc to look up specific nuances of the language. It should be more syntactical in nature, and less expository.

I didn't edit the whole topic with that goal in mind. I was mostly concerned about our duplication of content and missing Razor syntax content. Now that I've made some headway with those problems, I'll turn my attention broadly to the coverage with your feedback in mind. Let's see if I can improve it on the 2nd draft of updates.

> I'm concerned that some of the content we've moved into this doc is probably better left where it was.

When I moved things here, I thought that what I moved applied to both view/RP **_and_** component scenarios, and I tried to make the text and examples agnostic to that distinction. When I get into your feedback, I'll see which parts look like they should return to the *Components* topic.

guardrex commented on aspnetcore/mvc/views/razor.md:
````diff
@@ -512,92 +585,279 @@ If "rick@contoso.com" is passed in the model, the view generates the following H
+@namespace Your.Namespace.Here
+```
+
+If a page or view imports API with an `@namespace` directive, the original file's namespace is set relative to that namespace. 
````
I didn't write that. This ain't my topic. 😜 lol (... *but I'll see what I can do with it on the 2nd pass*)

guardrex commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -442,43 +442,113 @@ public class _Views_Something_cshtml : RazorPage<dynamic>
+::: moniker range=">= aspnetcore-3.0"
 
-### @model
+### \@code
```
> all-encompassing Razor

I asked about that. Dan said, https://github.com/aspnet/AspNetCore.Docs/issues/13465#issuecomment-514818607.

> differentiate cshtml and .razor pages

If you mean for code blocks, *yes*. That's the only effort that I'm aware of at this time. It isn't done yet tho. When it is, we'll be able to have `cshtml` and `razor` code blocks. For now, we can only use `cshtml`.

guardrex commented:
> I agree with @NTaylorMullen that we need to make it super clear in this doc which Razor language features are supported in .cshtml vs .razor. We should decide a mechanism to do that.

:question: For a given section that needs to call it out ...

*This scenario only applies to Razor components (.razor).*

*This scenario only applies to MVC views and Razor Pages (.cshtml).*

guardrex commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -369,13 +369,13 @@ In C#, a `using` statement is used to ensure an object is disposed. In Razor, th
 
 Scope-level actions can be performed with [Tag Helpers](xref:mvc/views/tag-helpers/intro).
```
As part of the effort to streamline this topic, I'll strike this line. Generally, I'll try to make each section's content pertain to the gesture at hand.

guardrex commented on aspnetcore/mvc/views/razor.md:
````diff
@@ -512,92 +585,279 @@ If "rick@contoso.com" is passed in the model, the view generates the following H
+<input maxlength="10" placeholder="Placeholder text" required="required" size="50">
+```
+
+### \@bind
````
Given that I didn't know that the directive attributes only apply to components, the 2nd draft commit will revert back to coverage in the *Components* topic only, which was approved by engineering when it was written. Of course, we can take a 2nd look at that if you wish.

guardrex commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -512,92 +585,279 @@ If "rick@contoso.com" is passed in the model, the view generates the following H
 
-If a page or view imports API with an `@namespace` directive, the original file's namespace is set relative to that namespace. 
+> [!NOTE]
+> Do **not** use the loop variable (`i`) in a `for` loop directly in a lambda expression. Otherwise the same variable is used by all lambda expressions causing `i`'s value to be the same in all lambdas. Always capture its value in a local variable (`buttonNumber` in the preceding example) and then use it.
```
When searching, it's only immediately evident in a VB topic ...

https://docs.microsoft.com/dotnet/visual-basic/language-reference/error-messages/using-the-iteration-variable-in-a-lambda-expression-may-have-unexpected-results

... which probably explains why I didn't cross-link it.

However, there might be something over there that didn't come up in my searches. @BillWagner, do you know if we have C# coverage similar to that :point_up: VB coverage?

guardrex commented:
Don't look :eyes: yet ... the last commit was just 🔔 **_Round 1_** 🥊. Tomorrow (Sunday), I'll step out and address the broader feedback on tightening up the language throughout the *Razor syntax* topic.

I'll ping ya when it's ready for another :eye:.

guardrex commented:
@NTaylorMullen @danroth27 Ready for another :eye: ... I'll number these points to make it easy to respond ...

1. I *think* that I have the bits split better now between the *Razor syntax* and *Components*.
1. The call-outs for 'this is MVC/RP only' and 'this is for RC only' are here. No flourishes (graphics), but we can ask docs team management if you would like to pursue it.
1. Directive attribute content (e.g., `@bind`) has been reviewed in the *Components* topic, so I think we're ok there.
1. Didn't link the VB topic for [loop directly in a lambda expression](https://docs.microsoft.com/dotnet/visual-basic/language-reference/error-messages/using-the-iteration-variable-in-a-lambda-expression-may-have-unexpected-results). Can't find a C# equivalent. Left that as-is for the time being.
1. **"blocks"** ... how shall I proceed? ...

   * ... 'in the `@( ... )` expression' 👍
   * ... 'in the `@{ ... }` code block' 👍
   * ... 'in the `@functions` directive' or 'in the `@functions` code block' (or drop 'code')?
   * ... 'in the `@code` directive' or 'in the `@code` block'?

1. For ...

   > the intent is that the Razor syntax reference doc should remain a syntax reference, not an entry point for learning the language.

   I read the whole topic, it seems ok. If you see something that you want to 🔪, I'll get to hack'in it out. :ear:

NTaylorMullen commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -442,43 +440,113 @@ public class _Views_Something_cshtml : RazorPage<dynamic>
+::: moniker range=">= aspnetcore-3.0"
 
-### @model
+### \@code
```
I'd add the this is only applicable in .razor files snippet

NTaylorMullen commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -442,43 +440,113 @@ public class _Views_Something_cshtml : RazorPage<dynamic>
-The `@model` directive specifies the type of this property. The directive specifies the `T` in `RazorPage<T>` that the generated class that the view derives from. If the `@model` directive isn't specified, the `Model` property is of type `dynamic`. The value of the model is passed from the controller to the view. For more information, see [Strongly typed models and the &commat;model keyword](xref ...
+::: moniker-end
+
+### \@implements
```
This was added in 3.0, also it uses an `@code` in the body. I'd change it to `@functions` so it's .cshtml/.razor compatible.

guardrex commented:
@danroth27 No giant rush, but this is blocking a couple of Blazor issues. I'm going to rebase the staging branch after we merge this and pick up with localization and globalization Blazor issues.

... going for a no-:boom: no-:cry: experience for Scott when he merges the staging branch later.

danroth27 commented on aspnetcore/blazor/common/samples/3.x/BlazorSample/README.md:
```diff
@@ -4,7 +4,7 @@ This sample illustrates the use of Blazor scenarios described in the Blazor docu
 ## Call web API example
 
-The web API example requires a running web API based on the sample app for the <a href="https://docs.microsoft.com/aspnet/core/tutorials/first-web-api">Tutorial: Create a web API with ASP.NET Core MVC</a> topic. The sample app makes requests to the web API at `https://localhost:10000/api/todo`. If a different web API address is used, update the `ServiceEndpoint` constant value in the Razor compon ...
+The web API example requires a running web API based on the sample app for the <a href="https://docs.microsoft.com/aspnet/core/tutorials/first-web-api">Tutorial: Create a web API with ASP.NET Core MVC</a> topic. The sample app makes requests to the web API at `https://localhost:10000/api/todo`. If a different web API address is used, update the `ServiceEndpoint` constant value in the Razor compon ...
```
OK, I spoke to @rynowak about this and he is fine with the phrase "@code block" or "@functions block". His main concern is to avoid calling the block a "code block", since it contains members not arbitrary code. That's what `@{ }` is for. So we should revert this block->directive changes in this PR.

guardrex commented:
Thanks @danroth27 ... I'll go through now and revert the "block" changes.

If we manage to get this merged later this evening, then it will unblock my next round of work.

guardrex commented:
@danroth27 Ok ... I think that licked 😋 'um.

danroth27 commented on aspnetcore/blazor/components.md:
````diff
@@ -366,7 +366,15 @@ In the following example, `UpdateHeading` is called asynchronously when the butt
 ```
 
-For some events, event-specific event argument types are permitted. If access to one of these event types isn't necessary, it isn't required in the method call.
+Lambda expressions can also be used:
````
This text seems to be duplicated from below (line 399)

danroth27 commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -442,43 +440,111 @@ public class _Views_Something_cshtml : RazorPage<dynamic>
+
+### \@functions
+
+The `@functions` directive enables a Razor Page to add C# members (fields, properties, and methods) to a view:
```
```suggestion
The `@functions` directive enables adding C# members (fields, properties, and methods) to the generate class:
```

danroth27 commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -512,92 +581,157 @@ If "rick@contoso.com" is passed in the model, the view generates the following H
+
+The `@page` directive has different effects depending on the type of the file where it appears. The directive:
+
+* Turns a file into an MVC action, which means that the file handles requests directly, without going through a controller. For more information, see <xref:razor-pages/index>.
```
While this is technically true, but I think it would be simpler to say that the `@page` directive in a .cshtml file indicates that the file is a Razor Page. Is there a reason this wording was used?

guardrex replied:
Taken from engineering language ... sounds kind'a *Ryan-y* to me. :smile: lol

I'll use your language.

guardrex commented:
Thanks @danroth27 ... after I address anything that Rick turns up, I'll be moving on to globalization/localization issues this weekend. I hope to have that up by Monday morning.

Rick-Anderson commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -512,92 +581,157 @@ If "rick@contoso.com" is passed in the model, the view generates the following H
 ::: moniker-end
 
-### @attribute
+::: moniker range="< aspnetcore-3.0"
```
Why does this apply to < vers 3 only?

guardrex replied:
Can't split the bullets of a list. For >=3.0, there are two bullets (Razor Pages and Razor Components). For <3.0, there's only a single item in sentence format for Razor Pages. WRT "sentence format," I try to steer clear of single-bullet lists outside of *Additional Resources* sections.

Rick-Anderson commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -512,92 +581,157 @@ If "rick@contoso.com" is passed in the model, the view generates the following H
+::: moniker range="< aspnetcore-3.0"
 
-The `@attribute` directive adds the given attribute to the class of the generated page or view. The following example adds the `[Authorize]` attribute:
+The `@page` directive turns a file into an MVC action, which means that the file handles requests directly, without going through a controller. For more information, see <xref:razor-pages/index>.
```
The `@page` directive on the first line of a *.cshtml* file indicates that the file is a [Razor Page](xref:razor-pages/index).

Rick-Anderson commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -512,92 +581,157 @@ If "rick@contoso.com" is passed in the model, the view generates the following H
+*This scenario only applies to Razor components (.razor).*
 
-If a page or view imports API with an `@namespace` directive, the original file's namespace is set relative to that namespace. 
+A component can capture and render additional attributes in addition to the component's declared parameters using the `@attributes` Razor directive. For more information, see <xref:blazor/components#attribute-splatting-and-arbitrary-parameters>.
```
`@attributes`  allows a component to render non-declared attributes. For more information, see <xref:blazor/components#attribute-splatting-and-arbitrary-parameters>.

Rick-Anderson commented on aspnetcore/mvc/views/razor.md:
```diff
@@ -442,43 +440,111 @@ public class _Views_Something_cshtml : RazorPage<dynamic>
+### \@code
 
-The `@model` directive specifies the type of the model passed to a view:
+*This scenario only applies to Razor components (.razor).*
```
Create a new H2 at the end of the file 

## Razor components specific

This scenario only applies to Razor components (.razor)

Move all 7 of your `This scenario only applies to Razor components (.razor)` to that section.

guardrex replied:
Ping @danroth27 @NTaylorMullen for more info here on future plans. (You can skip to the last line if you want. :wink:)

I'm concerned about the layout: There are *directives* and *directive attributes*, so the topic might end up with four sections depending on what the framework offers down the road if we split out RC right now:

* RP/views
  * Directives
  * Directive attributes
* RC
  * Directives
  * Directive attributes

I was under the impression that at least some of the Razor Component-only scenarios (directive attributes) would be available for RP/views one day. Therefore, I structured the content to handle those framework enhancements should they occur ...

* Directives (alphabetical with qualifying lines for RP/view-only and RC-only scenarios)
* Directive attributes (alphabetical with qualifying lines for RP/view-only and RC-only scenarios)

The current layout provides for straightforward updates ...

* Just remove the 'This is RC only' line for any of the scenarios added on this PR when one or more become available for RP/views in the future.
* In the future, a scenario is added by the framework+docs to either the *Directives* or *Directive attributes* sections and ...
  * No qualifying line indicates that the scenario applies to both.
  * Use the 'This is RP/views only' line.
  * Use 'This is RC only' line.

Couldn't be simpler imo ... for the readers and the team.

If we go with a RC-specific section and some of the RC-only scenarios are extended to cover RP/views, duplicate content will result or the layout will need to go to what I have on this PR.

**TL;DR** ... *the 9-volume book!* :smile: 

**Will some or all of these new Razor Components Razor syntax scenarios (perhaps just the directive attributes) be coming to a Razor Pages page or MVC view _near you?!_** :point_left: If **_No! Never!_**, then let's try the RC-only section approach (but I'm also concerned about the 'end of the file' location ... that seems out-of-order). If **_Sure thing! Why the heck not?!_** :tada:, then I recommend the current approach on the PR now.

danroth27 commented:
> Will some or all of these new Razor Components Razor syntax scenarios (perhaps just the directive attributes) be coming to a Razor Pages page or MVC view near you?!

Yes, it's certainly possible that this might happen. 

The Razor Syntax doc is referential, where the individual section will often be referenced directly from other documents. For this reason I think it's better that each section to carry the note whether it applies generally or just to .cshtml/.razor.

guardrex commented:
Cool ... I recommend that we stick with the current layout.

guardrex commented:
I'll make the final round of updates this afternoon.
